community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala

```diff
@@ -371,104 +421,189 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
     graph.createConstraint("Person","ssn")
 
     // when
-    val result = execute("MERGE (person:Person {ssn:{p}.ssn}) ON CREATE SET person = {p} RETURN person",
-      "p"->Map("ssn" -> 42, "name"->"Robert Paulsen"))
+    val result = updateWithBothPlanners("MERGE (person:Person {ssn:{p}.ssn}) ON CREATE SET person = {p} RETURN person.ssn",
+      "p" -> Map("ssn" -> 42, "name"->"Robert Paulsen"))
 
     // then - does not throw
-    assertStats(result, nodesCreated = 1, labelsAdded = 1, propertiesSet = 3/*should really be 2!*/)
+    result.toList should equal(List(Map("person.ssn" -> 42)))
+    assertStats(result, nodesCreated = 1, labelsAdded = 1, propertiesSet = 3)
   }
 
-  test("should_work_when_finding_multiple_elements") {
-    assertStats(execute( "CREATE (:X) CREATE (:X) MERGE (:X)"), nodesCreated = 2, labelsAdded = 2)
+  test("should work when finding multiple elements") {
+    assertStats(updateWithBothPlanners( "CREATE (:X) CREATE (:X) MERGE (:X)"), nodesCreated = 2, labelsAdded = 2)
   }
 
-  test("should_support_updates_while_merging") {
+  test("merge should handle argument properly") {
+    createNode("x" -> 42)
+
+    val query = """WITH 42 AS x MERGE (c:N {x: x})"""
+
+    val result = updateWithBothPlanners(query)
+
+    assertStats(result, nodesCreated = 1, labelsAdded = 1, propertiesSet = 1)
+  }
+
+  test("should be able to merge using property from match") {
+    createLabeledNode(Map("name" -> "A", "bornIn" -> "New York"), "Person")
+    createLabeledNode(Map("name" -> "B", "bornIn" -> "Ohio"), "Person")
+    createLabeledNode(Map("name" -> "C", "bornIn" -> "New Jersey"), "Person")
+    createLabeledNode(Map("name" -> "D", "bornIn" -> "New York"), "Person")
+    createLabeledNode(Map("name" -> "E", "bornIn" -> "Ohio"), "Person")
+    createLabeledNode(Map("name" -> "F", "bornIn" -> "New Jersey"), "Person")
+
+    val query = "MATCH (person:Person) MERGE (city: City {name: person.bornIn}) RETURN person.name"
+
+    val result = updateWithBothPlanners(query)
+
+    assertStats(result, nodesCreated = 3, propertiesSet = 3, labelsAdded = 3)
+  }
+
+  test("should be able to merge using property from match with index") {
+    graph.createIndex("City", "name")
+
+    createLabeledNode(Map("name" -> "A", "bornIn" -> "New York"), "Person")
+    createLabeledNode(Map("name" -> "B", "bornIn" -> "Ohio"), "Person")
+    createLabeledNode(Map("name" -> "C", "bornIn" -> "New Jersey"), "Person")
+    createLabeledNode(Map("name" -> "D", "bornIn" -> "New York"), "Person")
+    createLabeledNode(Map("name" -> "E", "bornIn" -> "Ohio"), "Person")
+    createLabeledNode(Map("name" -> "F", "bornIn" -> "New Jersey"), "Person")
+
+    val query = "MATCH (person:Person) MERGE (city: City {name: person.bornIn}) RETURN person.name"
+
+    val result = updateWithBothPlanners(query)
+
+    assertStats(result, nodesCreated = 3, propertiesSet = 3, labelsAdded = 3)
+  }
+
+  test("should be able to use properties from match in ON CREATE") {
```

Comment:
We are lacking tests which have queries using both `ON MATCH` and `ON CREATE`. We should add some.